WalletWasabi.Tests/UnitTests/BitcoinCore/MempoolTests.cs

```diff
@@ -0,0 +1,153 @@
+using NBitcoin;
+using System.Collections.Generic;
+using System.Linq;
+using WalletWasabi.BitcoinCore.Mempool;
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests.BitcoinCore;
+
+/// <summary>
+/// Tests for <see cref="Mempool"/>.
+/// </summary>
+public class MempoolTests
+{
+	/// <summary>
+	/// Verifies behavior for <see cref="Mempool.AddTransaction(Transaction)"/>.
+	/// </summary>
+	[Fact]
+	public void AddRemoveTransaction()
+	{
+		// Create the tx1 transaction.
+		Transaction tx1;
+		OutPoint tx1prevOut1;
+		{
+			tx1 = Network.Main.CreateTransaction();
+			tx1prevOut1 = new(hashIn: new uint256(0x1), nIn: 0);
+			tx1.Inputs.Add(new TxIn(tx1prevOut1));
+
+			OutPoint tx1prevOut2 = new(hashIn: new uint256(0x2), nIn: 0);
+			tx1.Inputs.Add(new TxIn(tx1prevOut2));
+
+			TxOut tx1Out1 = new(Money.Coins(0.1m), scriptPubKey: Script.Empty);
+			tx1.Outputs.Add(tx1Out1);
+		}
+
+		// Create the tx2 transaction that spends the same tx1prevOut1 as tx1.
+		Transaction tx2;
+		{
+			tx2 = Network.Main.CreateTransaction();
+			tx2.Inputs.Add(new TxIn(tx1prevOut1));
+
+			OutPoint tx2prevOut2 = new(hashIn: new uint256(0x3), nIn: 0);
+			tx2.Inputs.Add(new TxIn(tx2prevOut2));
+
+			TxOut tx2Out1 = new(Money.Coins(0.1m), scriptPubKey: Script.Empty);
+			tx1.Outputs.Add(tx2Out1);
+		}
+
+		Mempool mempool = new();
+
+		Assert.False(mempool.ContainsTransaction(tx1.GetHash()));
+
+		mempool.AddTransaction(tx1);
+		Assert.True(mempool.ContainsTransaction(tx1.GetHash()));
+
+		// Tests that no exception is thrown.
+		mempool.AddTransaction(tx1);
+		Assert.True(mempool.ContainsTransaction(tx1.GetHash()));
+		Assert.Equal(new HashSet<uint256>() { tx1.GetHash() }, mempool.GetMempoolTxids());
+
+		// Now we expect that tx1 will be removed from the mempool because it spends
+		// the same transaction output as tx2.
+		mempool.AddTransaction(tx2);
+		Assert.False(mempool.ContainsTransaction(tx1.GetHash()));
+		Assert.True(mempool.ContainsTransaction(tx2.GetHash()));
+		Assert.Equal(new HashSet<uint256>() { tx2.GetHash() }, mempool.GetMempoolTxids());
```

Comment:
FTR: I know this respects previously existing behaviour but I cannot ignore the fact that it doesn't distinguish between replaceable transactions and double spending. And even if we consider it a valid policy as full RBF this still fail to validate the new transaction pays more fee than the previous one.

Reply:
Note: once you remove a transaction you should remove all the descendant transactions too.